src/ActivityLogger.php

```diff
@@ -2,129 +2,123 @@
 
 namespace Spatie\Activitylog;
 
-use Illuminate\Auth\AuthManager;
 use Illuminate\Contracts\Config\Repository;
 use Illuminate\Database\Eloquent\Model;
 use Illuminate\Support\Arr;
 use Illuminate\Support\Carbon;
 use Illuminate\Support\Str;
 use Illuminate\Support\Traits\Macroable;
 use Spatie\Activitylog\Contracts\Activity as ActivityContract;
-use Spatie\Activitylog\Exceptions\CouldNotLogActivity;
 
 class ActivityLogger
 {
     use Macroable;
 
-    /** @var \Illuminate\Auth\AuthManager */
-    protected $auth;
+    protected string $defaultLogName = '';
 
-    protected $defaultLogName = '';
+    protected CauserResolver $causerResolver;
 
-    /** @var string */
-    protected $authDriver;
+    protected ActivityLogStatus $logStatus;
 
-    /** @var \Spatie\Activitylog\ActivityLogStatus */
-    protected $logStatus;
+    protected ?ActivityContract $activity = null;
 
-    /** @var \Spatie\Activitylog\Contracts\Activity */
-    protected $activity;
+    protected ActivityLoggerBatch $batch;
 
-    public function __construct(AuthManager $auth, Repository $config, ActivityLogStatus $logStatus)
+    public function __construct(Repository $config, ActivityLogStatus $logStatus, ActivityLoggerBatch $batch, CauserResolver $causerResolver)
     {
-        $this->auth = $auth;
+        $this->causerResolver = $causerResolver;
 
-        $this->authDriver = $config['activitylog']['default_auth_driver'] ?? $auth->getDefaultDriver();
+        $this->batch = $batch;
 
         $this->defaultLogName = $config['activitylog']['default_log_name'];
 
         $this->logStatus = $logStatus;
     }
 
-    public function setLogStatus(ActivityLogStatus $logStatus)
+    public function setLogStatus(ActivityLogStatus $logStatus): self
     {
         $this->logStatus = $logStatus;
 
         return $this;
     }
 
-    public function performedOn(Model $model)
+    public function performedOn(Model $model): self
     {
         $this->getActivity()->subject()->associate($model);
 
         return $this;
     }
 
-    public function on(Model $model)
+    public function on(Model $model): self
     {
         return $this->performedOn($model);
     }
 
-    public function causedBy($modelOrId)
+    public function causedBy(mixed $modelOrId): self
```

Comment:
We can type-hint what the `resolve()` method accepts. Even if that's nearly every scalar - but `bool` and `array` is simply impossible and `associate()` is limited to `\Illuminate\Database\Eloquent\Model|int|string` - I'm not sure about the `Closure` here!? 🤔
What would be the usecase or scenario to put in a Closure?

```suggestion
    public function causedBy(Model | int | string | null $modelOrId): self
```

Reply:
It just a convenient on the `causedBy()` method to just pass a clouser that will return whatever is acceptable as a cuser instated of explicitly define a closure on the resolve method first.

I'll add a test `it_can_log_an_activity_with_a_causer_resolved_from_closure`

```php
    /** @test */
    public function it_can_log_an_activity_with_a_causer_resolved_from_closure()
    {
        activity()
            ->causedBy(fn() => User::first())
            ->log($this->activityDescription);

        $firstActivity = Activity::first();

        $this->assertEquals(User::first()->id, $firstActivity->causer->id);
        $this->assertInstanceOf(User::class, $firstActivity->causer);
    }
```

Reply:
No, this is too hard to get and the `activity()` helper or underlying class is by far not responsible to set a resolver.
It's also not really understandable as this will be the resolver for all following activities.

```php
activity()->by(fn() => User::first())->log('foo');
activity()->log('bar');
```

The `bar` log will have the first user and not the current one as cause.
Who wants to customize the resolver logic should use the proper class/facade and method to do so.

Reply:
In this case you are right!

I never thought about it from this point of view!